Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift

```diff
@@ -1257,100 +1333,128 @@ private final class PubGrubPackageContainer {
+
+        var upperBounds: [PackageReference: Version] = [:]
+        self.queue.async(group: sync) {
+            upperBounds = compute(AnyCollection(versions.dropFirst(idx + 1)), upperBound: true)
+        }
+
+        var lowerBounds: [PackageReference: Version] = [:]
+        self.queue.async(group: sync) {
+            lowerBounds = compute(AnyCollection(versions.dropLast(versions.count - idx).reversed()), upperBound: false)
+        }
+
+        // FIXME: should this timeout be configurable?
+        switch sync.wait(timeout: .now() + 60) {
+        case .timedOut:
+            throw StringError("timeout computing \(node.package.name) bounds")
+        case .success:
+            return (lowerBounds, upperBounds)
+        }
     }
 }
 
+// 👀 IMPORANT CHANGE
 /// An utility class around PackageContainerProvider that allows "prefetching" the containers
 /// in parallel. The basic idea is to kick off container fetching before starting the resolution
 /// by using the list of URLs from the Package.resolved file.
 private final class ContainerProvider {
     /// The actual package container provider.
-    let provider: PackageContainerProvider
+    private let provider: PackageContainerProvider
 
     /// Wheather to perform update (git fetch) on existing cloned repositories or not.
-    let skipUpdate: Bool
+    private let skipUpdate: Bool
 
     /// Reference to the pins store.
-    let pinsMap: PinsStore.PinsMap
+    private let pinsMap: PinsStore.PinsMap
+
+    /// Queue to run async operations on
+    private let queue: DispatchQueue
+
+    //// Store cached containers
+    private var containersCache = ThreadSafeKeyValueStore<PackageReference, PubGrubPackageContainer>()
+
+    //// Store prefetches synchronization
+    private var prefetches = ThreadSafeKeyValueStore<PackageReference, DispatchGroup>()
 
-    init(_ provider: PackageContainerProvider, skipUpdate: Bool, pinsMap: PinsStore.PinsMap) {
+    init(provider: PackageContainerProvider, queue: DispatchQueue, skipUpdate: Bool, pinsMap: PinsStore.PinsMap) {
         self.provider = provider
+        self.queue = queue
         self.skipUpdate = skipUpdate
         self.pinsMap = pinsMap
     }
 
-    /// Condition for container management structures.
-    private let fetchCondition = Condition()
-
-    /// The list of fetched containers.
-    private var _fetchedContainers: [PackageReference: Result<PubGrubPackageContainer, Error>] = [:]
-
-    /// The set of containers requested so far.
-    private var _prefetchingContainers: Set<PackageReference> = []
+    /// Get a cached container for the given identifier, asserting / throwing if not found.
+    func getCachedContainer(for package: PackageReference) throws -> PubGrubPackageContainer {
+        guard let container = self.containersCache[package] else {
+            let error = "container for \(package.name) expected to be cached"
+            assertionFailure()
+            throw InternalError(error)
```

Comment:
I think throwing internal errors is a good pattern.  I think we still have a few `fatalError()` calls in the code base, but we should really eradicate them since it’s never right to take down the client.

Reply:
(I realize this particular case isn’t replacing a `fatalError()` but I mean in general)

Reply:
I agree, and plan to do a follow up PR to clean many of the `fatalError` in the code base